### PR TITLE
ENH: implement nep 0015: merge multiarray and umath

### PR DESCRIPTION
**Pull request by mattip**
~This s a first look at how the implementation of "NEP 0015" will look.~ Only one test pattern needed adjustment.

There is now `multiarray.py` and `umath.py`, and a c-extension `_multiarray` comprising the code from the two previous c-extension modules.

The next step is to clean up the module initialization to avoid  `set_numeric_ops`

Edit: s/test/test pattern/

Edit: strike out first sentence

### CODE REVIEW

File: numpy/core/code_generators/generate_ufunc_api.py
Comment (eric-wieser) [inline numpy/core/code_generators/generate_ufunc_api.py]: Why are we trying to import a top-level module?

**Comment by mattip** [comment]
about `_import_array` - AFAICT it has a few reasons for existing:
- make sure `multiarray` is imported (useful internally before the modules are merged)
- make sure `_ARRAY_API` is initialized, which is also checked when importing `multiarray`
- perform runtime checks on byte order

The second test seems redundant. The first is now redundant in numpy since they are one-in-the-same module. The third can be moved to `_multiarray_umath` initialization. So `_import_array` becomes an alias for third-party users to `PyImport_ImportModule("numpy.core.multiarray")` and we do not need to call it internally. It can also move out of `generate_numpy_api.py` to one of the numpy headers.

Am I missing something?

File: numpy/core/code_generators/generate_ufunc_api.py
Comment (eric-wieser) [reply numpy/core/code_generators/generate_ufunc_api.py]: I don't have enough experience here to know if we want this method.

I'm just worried that this looks like a removed-in-py3k implicit relative import.

Comment (mattip) [reply numpy/core/code_generators/generate_ufunc_api.py]: removed. It was useful only to import `_multiarray` indepedent of the rest of numpy:
```
data = imp.find_module('_multiarray', ['cpython3_virt/lib/python3.5/site-packages/numpy/core'])
multiarray = imp.load_module('_multiarray', *data)
```
which I was doing for debugging the bootstrap of numpy.__init__

**Comment by mattip** [comment]
Still WIP. I did not adapt `umathmodule.c` for the `c-api` import as documented [here](https://docs.scipy.org/doc/numpy-1.14.0/reference/c-api.array.html#importing-the-api) so calling any c-api function from `umathmodule` crashes. I tried changing `PY_ARRAY_UNIQUE_SYMBOL` to `NO_IMPORT_ARRAY` but the initialization still crashes. Still investigating.

File: numpy/core/setup.py
Comment (mhvk) [inline numpy/core/setup.py]: `umath` part!

File: numpy/core/setup.py
Comment (mhvk) [inline numpy/core/setup.py]: These disappear because they are already included above, correct?

File: numpy/core/code_generators/generate_ufunc_api.py
Comment (mhvk) [inline numpy/core/code_generators/generate_ufunc_api.py]: I'm very confused, shouldn't it be `_multiarray_umath` here and below?

**Review by mhvk — COMMENTED** [review]
So far, it looks far less intrusive than I would have thought! Some mostly ignorant queries...

**Comment by charris** [comment]
Just to note that I want to branch 1.15 before this goes in.

File: numpy/core/code_generators/generate_ufunc_api.py
Comment (mattip) [reply numpy/core/code_generators/generate_ufunc_api.py]: fixed, thanks

File: numpy/core/setup.py
Comment (mattip) [reply numpy/core/setup.py]: correct

File: numpy/core/setup.py
Comment (mattip) [reply numpy/core/setup.py]: copy-paste. Fixed

**Comment by mhvk** [comment]
@mattip - I think it is very good to see it happening, especially as this is quite an involved PR. I quite like the approach of doing it as much as possible in small pieces. Anyway, I would not worry about the "noise" - you warned us amply that it was a WIP! Just ping when it is useful to look again.

**Comment by mattip** [comment]
Rebased off master after 1.15

**Comment by mhvk** [comment]
I'm quite excited about this - do ping if this is ready for further review!

**Comment by mattip** [comment]
I took the WIP out of the title and am ready for a review. While the NEP (PR #10704) itself has not progressed, I have implemented the minimal set of changes I could to merge the two c-extension modules. Merging this would be helpful for continuing the work on matmul (PR #11133) since it would give access to the `cblasfuncs.c` routines from umath. That is why I added commits a191224 and b997efd. 

Note that I only moved files that have not changed in quite a while and that have no `NUMPY_API` functions into `common` (used to be `private`) in order to minimize git churn on possible pull requests and issues. Eventually, we may wish to restructure the source code directories according to a different paradigm.

File: numpy/core/setup.py
Comment (mhvk) [inline numpy/core/setup.py]: Why the commented out files? Just remove? Or at least comment on why they are commented out... Is it because they are already listed in `multiarray`?

File: numpy/core/setup.py
Comment (mhvk) [inline numpy/core/setup.py]: Might it be an idea to have a `common_src` here?

File: numpy/core/src/multiarray/multiarraymodule.c
Comment (mhvk) [inline numpy/core/src/multiarray/multiarraymodule.c]: Might it be more logical to have this contained in a separate `umathmodule.h` file?

File: numpy/core/src/multiarray/multiarraymodule.c
Comment (mhvk) [inline numpy/core/src/multiarray/multiarraymodule.c]: If following the separate `umathmodule.h` file idea above, these lines could go in there too.

File: numpy/core/src/umath/umathmodule.c
Comment (mhvk) [inline numpy/core/src/umath/umathmodule.c]: Given the comment, should this be removed?

**Review by mhvk — COMMENTED** [review]
Looks very good, and I'm excited about it! My comments are really two, both organisational: whether to have `common_src` in `setup.py` and a new `umathmodule.h` to keep things umath-related in the multiarray initialization.

File: numpy/core/setup.py
Comment (mattip) [reply numpy/core/setup.py]: removed

File: numpy/core/setup.py
Comment (mattip) [reply numpy/core/setup.py]: adopted. Eventually everything will become common

File: numpy/core/src/umath/umathmodule.c
Comment (mattip) [reply numpy/core/src/umath/umathmodule.c]: yes

File: numpy/core/setup.py
Comment (mattip) [reply numpy/core/setup.py]: fixed

File: numpy/core/src/multiarray/multiarraymodule.c
Comment (mattip) [reply numpy/core/src/multiarray/multiarraymodule.c]: I cleaned this up by removing many uneeded include directives

File: numpy/core/src/multiarray/multiarraymodule.c
Comment (mattip) [reply numpy/core/src/multiarray/multiarraymodule.c]: cleanup up via removal and refactoring, without a umathmodule.h

Comment (mattip) [reply numpy/core/src/multiarray/multiarraymodule.c]: on second thought umathmodule.h would be better

**Comment by mattip** [comment]
rebased off master to solve merge conflicts, added common/umathmodule.h

**Comment by mattip** [comment]
It turns out this would help for matmul-as-ufunc. The `TYPE_dot` (`TYPE`=`FLOAT`, `DOUBLE`,...) functions sit in `arraytypes.c.src` which is in `multiarray`, the loops I am working on live in `umath/loops.c.src`

File: numpy/core/setup.py
Comment (mhvk) [inline numpy/core/setup.py]: clearly got to the nitpicking state: inconsistent indentation...

File: numpy/core/src/umath/umathmodule.c
Comment (mhvk) [inline numpy/core/src/umath/umathmodule.c]: I think this commented out section can now go, no?

**Review by mhvk — COMMENTED** [review]
Two small comments left, but I do feel somewhat out of my league here, so it would be good if someone else could give a final review as well. 

Do agree that it would be nice to merge it, otherwise you'll keep having to rebase...

File: numpy/core/setup.py
Comment (mattip) [reply numpy/core/setup.py]: fixed

File: numpy/core/src/umath/umathmodule.c
Comment (mattip) [reply numpy/core/src/umath/umathmodule.c]: removed, also the next comment and an unneeded include

**Comment by mattip** [comment]
This is ready for review

**Comment by mhvk** [comment]
As I noted above, I'm quite happy with this, but think it would be best if someone else reviewed as well, as I do feel somewhat out of my league. @eric-wieser had another look earlier; others might be @charris or @njsmith.

**Review by tylerjereddy — COMMENTED** [review]
I added some minor comments from a first pass after reading the NEP and the changes here.

It is certainly a positive indicator that all CI tests pass with almost completely negligible unit test adjustments.

As noted in a previous comment in the PR, this implements only the first of the 3 proposed changes in the NEP, which is probably sensible given the potential invasiveness of deprecating / replacing `set_numeric_ops`, which I suppose touches a plethora of `ndarray` (operator) `ndarray` situations. 

Does this reorganization provide any additionally testable benefits? I think what I'm asking here is if you can test at the Python level that imports are less messy / circular in some tangible way compared to before this PR branch?

For example, `len(dir(np))` is around 622 at the moment--is there a Python object somewhere whose `dir` is a bit "cleaner" now or that should be tested for regression moving forward?

Conversely, the whole idea may be that this is simply a low-level reorganization that effectively perfectly preserves exposed behavior at the Python level, so no new tests are really appropriate.

File: numpy/core/src/umath/umathmodule.c
Comment (mattip) [reply numpy/core/src/umath/umathmodule.c]: Returning -1 will trigger a python-level exception. the `goto error` will set a generic exception, but we can sometimes do better by passing on the exception set by other c-api functions and simply returning -1 immediately.

File: numpy/core/src/umath/umathmodule.c
Comment (mattip) [reply numpy/core/src/umath/umathmodule.c]: fixed

**Comment by mattip** [comment]
Indeed this PR does not change `set_numeric_ops`, but simply reorganizes the C code to create a single shared object. As stated in the NEP, the intention is to only deprecate `np.set_numeric_ops`, so no changes in `dir(np)` were proposed nor implemented.

**Comment by mattip** [comment]
Rebased off master to fix merge conflicts.

File: numpy/core/multiarray.py
Comment (eric-wieser) [inline numpy/core/multiarray.py]: I think this file could do with either a docstring or a comment, explaining why it exists and is just a bunch of imports

File: numpy/core/umath.py
Comment (eric-wieser) [inline numpy/core/umath.py]: A comment / docstring would be good in this file too.

**Comment by eric-wieser** [comment]
Generally looks pretty uncontroversial. Not super easy to review the series of one-line moves in the build tools, but tests passing seems a good indication of success.

The new `umath.py`  and `multiarray.py` look pretty ugly, but I'm not sure there's a better way to deal with them.

File: numpy/core/umath.py
Comment (mattip) [reply numpy/core/umath.py]: added to both

**Comment by mattip** [comment]
The codecov report now is clean for the merged PR, but the patch one seems to be erroneous, the lines in `numpy/ma/core.py` marked as newly-not-covered are not valid syntactically (one is a comment, others are `except:` and `else:` which must have been executed).

**Comment by tylerjereddy** [comment]
The [`diff` tab](https://codecov.io/gh/numpy/numpy/compare/e14db2f4fcbcf030701a0410a1d647edfa5f0261...2d34db0877ec8a4e79c6568893b325f8782dbf28/diff) shows 6 Python source files tracked by Codecov (C tracking not enabled yet). Confusingly, not sure how reliable the `changes` tab you are linking from is for the patch. There's a [docs page](https://docs.codecov.io/docs/unexpected-coverage-changes) that describes unexpected coverage changes & includes a support link for asking codecov staff if you really think something is wrong after dissecting the results files.

Coming back to the `diff` tab and parsing those results for Python source files:

-   `numpy/core/multiarray.py`: brand new file: +4 new lines covered
-   `numpy/lib/__init__.py`: holds steady at 100% coverage (+1 changed line is covered)
-  `numpy/core/setup.py`: 10 new lines of Python code that could have been covered were added (10 lines changed but not covered)
- `numpy/tests/test_ctypeslib.py `: no net change (adjusted 3 covered lines; adjusted 1 uncovered line)
- `numpy/core/numeric.py `: 4 covered lines were simplified to 1 covered line (we adjusted 4 lines but retain 1 line of coverage only)
- `numpy/core/umath.py `: add 4 covered lines in this brand new file (+4)
   
Total lines changed: ~27
Total lines covered: ~13
So about 50 % total coverage on the raw Python diff.

As @eric-wieser indicated previously, we may want to exclude build things like `setup.py` so the arithmetic is a little less silly. You may have some luck trying to contact the codecov team about the strange `changes` tab, but at least the `diff` tab looks sensible for generating the number they complain about.

**Comment by mattip** [comment]
rebasing to see if codcov now ignores setup.py

**Review by eric-wieser — APPROVED** [review]
Looks good to me now, but probably a good idea to get another review.

One thing I wonder is if we can create multiple module objects from within the extension module, rather than merging all the functions and splitting them again. In particular, we've now limited ourselves to not having `umath.x` and `multiarray.x` at the same time. Probably easier to investigate after dropping py2 support

**Comment by mattip** [comment]
@mhvk care to take another look?

**Comment by mhvk** [comment]
@mattip - as I noted above, I feel quite out of my league here, but with @eric-wieser agreeing, I'd say we're OK to go. Maybe a last ping to @charris, @njsmith, @pv - fine to merge in a few days unless they ask for a bit of time to have a look.

**Comment by charris** [comment]
I'd like to wait until 1.15.1 is out, which should be in about a week.

**Comment by mattip** [comment]
Rebased off master to see if gcov messages noted in #11790 go away after this PR. Also properly formatted commit messages and sqaushed a few together.

**Comment by tylerjereddy** [comment]
`mem_overlap.gcda` is still an issue here, but rebasing did resolve a few of the other timestamp issues; perhaps there are still components of the NumPy code base that drive redundant compilation beyond the changes in this PR then?

**Comment by mattip** [comment]
This seems a bit magical. Any ideas why it solves the problem? I don't mind making the change but perhaps we are simply masking a deeper issue. For instance, and this is just a theory I am making up, it may be that the gcov report for one extension module is getting silently written over by the other, where before the change at least we got the warning.

**Comment by mattip** [comment]
I would prefer the changes to the build order become a separate PR, built on top of this one.

**Comment by charris** [comment]
Let's get this in. Thanks @mattip  and the reviewers @eric-wieser , @mhvk , @tylerjereddy.

**Pull request merged**
